apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java

```diff
@@ -0,0 +1,265 @@
+        Arrays.asList(
+            new SimpleRecord(1, "a"),
+            new SimpleRecord(1, "b"),
+            new SimpleRecord(1, "c"),
+            new SimpleRecord(2, "d"),
+            new SimpleRecord(2, "e"));
+
+    ops.spark()
+        .createDataset(records, Encoders.bean(SimpleRecord.class))
+        .coalesce(1)
+        .writeTo(tableName)
+        .append();
+
+    // Create an equality delete manually since cannot force sql to do it
+    Table table = ops.getTable(tableName);
+    writeEqDeleteRecord(table, "data", "a");
+    writeEqDeleteRecord(table, "data", "d");
+    writeEqDeleteRecord(table, "data", "c");
+
+    // first run of compaction
+    RewriteDataFiles.Result result = rewriteFunc.apply(ops, table);
+    Assertions.assertEquals(1, result.addedDataFilesCount());
+    Assertions.assertEquals(1, result.rewrittenDataFilesCount());
+
+    // this asserts the dangling delete problem, since compaction did NOT remove the final delete
+    IcebergTableStats stats = ops.collectTableStats(tableName);
+    assertThat(stats.getNumEqualityDeleteFiles()).isEqualTo(3L);
+    assertThat(stats.getNumCurrentSnapshotEqualityDeleteFiles()).isEqualTo(1L);
+
+    // notice that running compaction again with no new commits cleans up the dangling delete
+    rewriteFunc.apply(ops, table);
+    stats = ops.collectTableStats(tableName);
+    assertThat(stats.getNumEqualityDeleteFiles()).isEqualTo(3L);
+    assertThat(stats.getNumCurrentSnapshotEqualityDeleteFiles()).isEqualTo(0L);
+  }
+
+  private void writeEqDeleteRecord(Table table, String delCol, Object delVal) {
+    List<Integer> equalityFieldIds = Lists.newArrayList(table.schema().findField(delCol).fieldId());
+    Schema eqDeleteRowSchema = table.schema().select(delCol);
+    Record record = GenericRecord.create(eqDeleteRowSchema).copy(ImmutableMap.of(delCol, delVal));
+    writeEqDeleteRecord(table, equalityFieldIds, eqDeleteRowSchema, record);
+  }
+
+  private void writeEqDeleteRecord(
+      Table table, List<Integer> equalityFieldIds, Schema eqDeleteRowSchema, Record deleteRecord) {
+    OutputFileFactory fileFactory =
+        OutputFileFactory.builderFor(table, 1, 1).format(FileFormat.ORC).build();
+    GenericAppenderFactory appenderFactory =
+        new GenericAppenderFactory(
+            table.schema(),
+            table.spec(),
+            ArrayUtil.toIntArray(equalityFieldIds),
+            eqDeleteRowSchema,
+            null);
+
+    EncryptedOutputFile file = createEncryptedOutputFile(null, fileFactory);
+
+    EqualityDeleteWriter<Record> eqDeleteWriter =
+        appenderFactory.newEqDeleteWriter(file, FileFormat.ORC, null);
+
+    try (EqualityDeleteWriter<Record> clsEqDeleteWriter = eqDeleteWriter) {
+      clsEqDeleteWriter.write(deleteRecord);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    table.newRowDelta().addDeletes(eqDeleteWriter.toDeleteFile()).commit();
+  }
+
+  private EncryptedOutputFile createEncryptedOutputFile(
```

Comment:
For my understanding: is encryption a related context here or this is mainly from Iceberg's testing code ?

Reply:
from iceberg's testing code,

i. tried to find a similar method without encryption since we don't need it and it is a bit of a distraction, but there doesn't seem to be one.